File: AlgorithmsAndStructures/GraphAlgorithms/TwoChinese.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

namespace AlgorithmsAndStructures.GraphAlgorithms
{
    class TwoChinese
    {
        private static Int64 GetMst(int root, List<Edge>[] graph)
        {
            Int64 result = 0;
            int n = graph.Length;
            int[] minInEdge = Enumerable.Repeat(int.MaxValue, n).ToArray();

            for (int from = 0; from < n; from++)
            {
                foreach (var edge in graph[from])
                {
                    minInEdge[edge.To] = Math.Min(minInEdge[edge.To], edge.Weight);
                }
            }

            for (int i = 0; i < n; i++)
                if (i != root)
                    result += (Int64)(minInEdge[i]);

            List<Edge>[] zeroGraph = new List<Edge>[n];
            for (int from = 0; from < n; from++)
            {
                zeroGraph[from] = new List<Edge>();
                foreach (var edge in graph[from])
                {
                    if (edge.Weight == minInEdge[edge.To])
                    {
                        edge.Weight = 0;
                        zeroGraph[from].Add(edge);
                    }
                }
            }

            if (CheckAvailability(root, zeroGraph))
                return result;

            int[] components = GetCondensation(zeroGraph);
            int componentsCount = components.Max() + 1;
            List<Edge>[] newGraph = new List<Edge>[componentsCount];
            for (int i = 0; i < componentsCount; i++)
            {
                newGraph[i] = new List<Edge>();
            }

            for (int from = 0; from < n; from++)
            {
                foreach (var edge in graph[from]
                .Where(edge => components[from] != components[edge.To]))
                {
                    newGraph[components[from]].Add(new Edge(components[edge.To], edge.Weight - minInEdge[edge.To]));
                }
            }

            result += GetMst(components[root], newGraph);
            return result;
        }

        private static bool CheckAvailability(int from, List<Edge>[] graph)
        {
            int n = graph.Length;
            bool[] visited = new bool[n];
            Queue<int> queue = new Queue<int>();
            queue.Enqueue(from);
            visited[from] = true;
            while (queue.Any())
            {
                from = queue.Dequeue();
                foreach (var edge in graph[from]
                .Where(edge => !visited[edge.To]))
                {
                    queue.Enqueue(edge.To);
                    visited[edge.To] = true;
                }
            }

            return visited.All(v => v);
        }

        private static List<int> GetTopSort(List<Edge>[] graph)
        {
            int n = graph.Length;
            bool[] visited = new bool[n];

            List<int> topSort = new List<int>(n);

            void Dfs(int from)
            {
                visited[from] = true;
                foreach (var edge in graph[from]
                .Where(edge => !visited[edge.To]))
                {
                    Dfs(edge.To);
                }

                topSort.Add(from);
            }

            for (int i = 0; i < n; i++)
            {
                if (!visited[i])
                {
                    Dfs(i);
                }
            }

            topSort.Reverse();
            return topSort;
        }

        private static int[] GetCondensation(List<Edge>[] graph)
        {
            int n = graph.Length;
            List<Edge>[] graphTransposed = GetTransposed(graph);
            List<int> order = GetTopSort(graph);
            bool[] visited = new bool[n];
            int componentCount = 0;
            int[] components = new int[n];

            void Dfs(int from)
            {
                visited[from] = true;
                components[from] = componentCount;
                foreach (var edge in graphTransposed[from]
                .Where(edge => !visited[edge.To]))
                {
                    Dfs(edge.To);
                }
            }

            foreach (var from in order)
            {
                if (!visited[from])
                {
                    componentCount++;
                    Dfs(from);
                }
            }

            return components.Select(c => c - 1).ToArray();
        }

        private static List<Edge>[] GetTransposed(List<Edge>[] graph)
        {
            int n = graph.Length;
            List<Edge>[] graphTransposed = new List<Edge>[n];
            for (int i = 0; i < n; i++)
            {
                graphTransposed[i] = new List<Edge>();
            }

            for (int from = 0; from < n; from++)
            {
                foreach (var edge in graph[from])
                {
                    graphTransposed[edge.To].Add(new Edge(from, edge.Weight));
                }
            }

            return graphTransposed;
        }

        private static void Program()
        {
            int n, m;
            var input = Console.ReadLine().Split().Select(c => int.Parse(c)).ToArray();
            n = input[0];
            m = input[1];

            List<Edge>[] graph = new List<Edge>[n];
            for (int i = 0; i < n; i++)
            {
                graph[i] = new List<Edge>();
            }

            for (int i = 0; i < m; i++)
            {
                input = Console.ReadLine().Split().Select(c => int.Parse(c)).ToArray();
                graph[input[0] - 1].Add(new Edge(input[1] - 1, input[2]));
            }

            if (CheckAvailability(0, graph))
            {
                Console.WriteLine("YES");
                Console.WriteLine(GetMst(0, graph));
            }
            else
            {
                Console.WriteLine("NO");
            }
        }

        public static void Solve()
        {
            var stackSize = 100000000;
            Thread thread = new Thread(new ThreadStart(Program), stackSize);
            thread.Start();
        }
    }

    internal class Edge
    {
        public int To { get; private set; }
        public int Weight { get; set; }

        public Edge(int to, int weight)
        {
            To = to;
            Weight = weight;
        }
    }
}

File: AlgorithmsAndStructures/GraphAlgorithms/TwoChinese.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <fstream>

using namespace std;

typedef long long int64;

struct Edge
{
	int64 to;
	mutable int64 weight;

	Edge(int64 to, int64 weight) : to(to), weight(weight) {}
	Edge() {}

	friend istream& operator>> (istream& is, Edge& edge)
	{
		is >> edge.to >> edge.weight;
		edge.to--;
		return is;
	}
};

typedef vector<vector<pair<int64, int64>>> edgesLists;


const int64 inf = 228228228228228;

int64 getMst(int64 root, const edgesLists& graph);
bool checkAvailability(int64 from, const edgesLists& graph);
void countVertex(int64 from, const edgesLists& graph, vector<bool>& visited);
void getCondensation(const edgesLists& graph, vector<int64>& components);
edgesLists getTransposed(const edgesLists& graph);
void markComponents(int64 from, const edgesLists& graph, vector<int64>& components, int64& component);
void getTopSort(const edgesLists& graph, vector<int64>& topSort);
void sortByTimeOut(int64 from, const edgesLists& graph, vector<bool>& visited, vector<int64>& topSort);

int64 getMst(int64 root, const edgesLists& graph)
{
	int64 result = 0;
	int64 n = graph.size();
	vector<int64> minInEdge = vector<int64>(n, inf);

	for (int64 from = 0; from < n; ++from)
	{
		for (auto edge : graph[from])
		{
			minInEdge[edge.first] = min(minInEdge[edge.first], edge.second);
		}
	}

	for (int64 i = 0; i < n; ++i)
		if (i != root)
			result += minInEdge[i];

	edgesLists zeroGraph = edgesLists(n);
	for (int64 from = 0; from < n; ++from)
	{
		for (auto edge : graph[from])
		{
			if (edge.second == minInEdge[edge.first])
			{
				edge.second = 0;
				zeroGraph[from].push_back(edge);
			}
		}
	}

	if (checkAvailability(root, zeroGraph))
		return result;

	vector<int64> components = vector<int64>(n);
	getCondensation(zeroGraph, components);
	int64 componentsCount = 0;
	for (auto component : components)
	{
		componentsCount = max(componentsCount, component);
	}
	componentsCount++;

	edgesLists newGraph = edgesLists(componentsCount);
	for (int64 from = 0; from < n; ++from)
	{
		for (auto edge : graph[from])
		{
			if (components[from] != components[edge.first])
				newGraph[components[from]].emplace_back(components[edge.first], edge.second - minInEdge[edge.first]);
		}
	}

	return result + getMst(components[root], newGraph);
}

bool checkAvailability(int64 from, const edgesLists& graph)
{
	int64 n = graph.size();
	vector<bool> visited = vector<bool>(n);
	countVertex(from, graph, visited);
	for (auto v : visited)
		if (!v)
			return false;
	return true;
}

void countVertex(int64 from, const edgesLists& graph, vector<bool>& visited)
{
	visited[from] = true;
	for (auto edge : graph[from])
		if (!visited[edge.first])
			countVertex(edge.first, graph, visited);
}

void getCondensation(const edgesLists& graph, vector<int64>& components)
{
	int64 n = graph.size();
	edgesLists graphTransposed = edgesLists(n);
	for (int64 from = 0; from < n; ++from)
		for (auto edge : graph[from])
			graphTransposed[edge.first].emplace_back(from, edge.second);
	vector<int64> order;
	getTopSort(graph, order);
	int64 componentsCount = 0;

	for (auto from : order)
	{
		if (components[from] == 0)
			markComponents(from, graphTransposed, components, ++componentsCount);
	}

	for (auto& component : components)
	{
		component--;
	}
}

edgesLists getTransposed(const edgesLists& graph)
{
	int64 n = graph.size();
	edgesLists graphTransposed = edgesLists(n);

	for (int64 from = 0; from < n; ++from)
		for (auto edge : graph[from])
			graphTransposed[edge.first].emplace_back(from, edge.second);

	return graphTransposed;
}

void getTopSort(const edgesLists& graph, vector<int64>& topSort)
{
	int64 n = graph.size();
	vector<bool> visited = vector<bool>(n);

	for (int64 from = 0; from < n; ++from)
	{
		if (!visited[from])
			sortByTimeOut(from, graph, visited, topSort);
	}

	reverse(topSort.begin(), topSort.end());
}

void sortByTimeOut(int64 from, const edgesLists& graph, vector<bool>& visited, vector<int64>& topSort)
{
	visited[from] = true;
	for (auto& edge : graph[from])
		if (!visited[edge.first])
			sortByTimeOut(edge.first, graph, visited, topSort);
	topSort.push_back(from);
}

void markComponents(int64 from, const edgesLists& graph, vector<int64>& components, int64& component)
{
	components[from] = component;
	for (auto& edge : graph[from])
		if (components[edge.first] == 0)
			markComponents(edge.first, graph, components, component);
}



int main()
{
	ifstream fin("chinese.in");
	ofstream fout("chinese.out");
	int64 n, m;
	fin >> n;
	fin >> m;

	edgesLists graph = edgesLists(n);
	int64 from;
	pair<int64, int64> edge;
	for (int64 i = 0; i < m; ++i)
	{
		fin >> from >> edge.first >> edge.second;
		graph[from - 1].emplace_back(edge.first - 1, edge.second);
	}

	if (checkAvailability(0, graph))
	{
		fout << "YES\n";
		fout << getMst(0, graph) << "\n";
	}
	else
	{
		fout << "NO\n";
	}
}